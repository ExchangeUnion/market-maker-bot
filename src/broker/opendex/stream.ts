import { Stream } from '../stream';

class OpenDexStream extends Stream {
  public start = () => {
    return Promise.resolve();
  }
  public close = () => {
    return Promise.resolve();
  }
}

export { OpenDexStream };
