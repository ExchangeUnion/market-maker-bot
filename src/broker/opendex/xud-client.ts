import fs from 'fs';
import {
  credentials,
  ClientReadableStream,
  ServiceError,
} from '@grpc/grpc-js';
import { XudClient } from './proto/xudrpc_grpc_pb';
import {
  SwapSuccess,
  SubscribeSwapsRequest,
  GetBalanceRequest,
  GetBalanceResponse,
  PlaceOrderRequest,
  PlaceOrderResponse,
  RemoveOrderRequest,
  RemoveOrderResponse,
  TradingLimitsRequest,
  TradingLimitsResponse,
  OrderSide as XudOrderSide,
 } from './proto/xudrpc_pb';

interface GrpcResponse {
  toObject: () => unknown;
}

type XudOrderRequest = {
  orderId: string,
  price: number | string,
  quantity: number,
  pairId: string,
  orderSide: XudOrderSide,
};

class XudGrpcClient {
  private tlscertpath: string;
  private rpchost: string;
  private rpcport: number;
  private client!: XudClient;

  constructor(
    { tlscertpath, rpchost, rpcport }:
    {
      tlscertpath: string,
      rpchost: string,
      rpcport: number,
    },
  ) {
    this.tlscertpath = tlscertpath;
    this.rpchost = rpchost;
    this.rpcport = rpcport;
  }

  public start = () => {
    const cert = fs.readFileSync(this.tlscertpath);
    const sslCredentials = credentials.createSsl(cert);
    const options = {
      'grpc.ssl_target_name_override' : 'localhost',
      'grpc.default_authority': 'localhost'
    };
    this.client = new XudClient(
      `${this.rpchost}:${this.rpcport}`,
      sslCredentials,
      options,
    );
  }

  public subscribeSwaps = (): ClientReadableStream<SwapSuccess> => {
    const swapsRequest = new SubscribeSwapsRequest();
    swapsRequest.setIncludeTaker(true);
    return this.client.subscribeSwaps(swapsRequest);
  }

  public getBalance = async (currency?: string): Promise<GetBalanceResponse.AsObject> => {
    const request = new GetBalanceRequest();
    if (currency) {
      request.setCurrency(currency.toUpperCase());
    }
    const balances = await new Promise((resolve, reject) => {
      this.client.getBalance(request, this.processResponse(resolve, reject));
    });
    return balances as GetBalanceResponse.AsObject;
  }

  public tradingLimits = async (): Promise<TradingLimitsResponse.AsObject> => {
    const request = new TradingLimitsRequest();
    const limits = await new Promise((resolve, reject) => {
      this.client.tradingLimits(request, this.processResponse(resolve, reject));
    });
    return limits as TradingLimitsResponse.AsObject;
  }

  public newOrder = async (orderRequest: XudOrderRequest): Promise<PlaceOrderResponse.AsObject> => {
    if (typeof orderRequest.price === 'string') {
      delete orderRequest.price;
    }
    const request = new PlaceOrderRequest();
    request.setQuantity(orderRequest.quantity);
    request.setSide(orderRequest.orderSide);
    request.setPairId(orderRequest.pairId);
    if (orderRequest.price) {
      request.setPrice(orderRequest.price as number);
    }
    request.setOrderId(orderRequest.orderId);
    const placeOrderResponse = await new Promise((resolve, reject) => {
      this.client.placeOrderSync(request, this.processResponse(resolve, reject));
    });
    return placeOrderResponse as PlaceOrderResponse.AsObject;
  }

  public removeOrder = async (orderId: string) => {
    const request = new RemoveOrderRequest();
    request.setOrderId(orderId);
    const removeOrderResponseObj = await new Promise((resolve, reject) => {
      this.client.removeOrder(request, this.processResponse(resolve, reject));
    });
    return removeOrderResponseObj as RemoveOrderResponse.AsObject;
  }

  private processResponse = (resolve: (value: unknown) => unknown, reject: (value: unknown) => unknown) => {
    return (error: ServiceError | null, response: GrpcResponse) => {
      if (error) {
        reject(error);
      } else {
        const responseObj = response.toObject();
        resolve(responseObj);
      }
    };
  }

}

export { XudGrpcClient };
