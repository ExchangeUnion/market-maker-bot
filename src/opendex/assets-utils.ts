import { BigNumber } from 'bignumber.js';
import { Logger } from '../logger';
import { GetBalanceResponse, TradingLimitsResponse } from '../proto/xudrpc_pb';
import { OpenDEXassetAllocation } from '../trade/info';
import { satsToCoinsStr } from '../utils';
import { errors } from './errors';

type LogAssetBalanceParams = {
  logger: Logger;
  assets: OpenDEXassetAllocation;
};

const logAssetBalance = ({ logger, assets }: LogAssetBalanceParams): void => {
  const {
    baseAssetBalance,
    baseAssetMaxInbound,
    baseAssetMaxOutbound,
    quoteAssetBalance,
    quoteAssetMaxInbound,
    quoteAssetMaxOutbound,
  } = assets;
  logger.trace(
    `Base asset balance ${baseAssetBalance.toFixed()} (maxinbound: ${baseAssetMaxInbound.toFixed()}, maxoutbound: ${baseAssetMaxOutbound.toFixed()}) and quote asset balance ${quoteAssetBalance.toFixed()} (maxinbound: ${quoteAssetMaxInbound.toFixed()}, maxoutbound: ${quoteAssetMaxOutbound.toFixed()}).`
  );
};

type ParseOpenDEXassetsParams = {
  balanceResponse: GetBalanceResponse;
  tradingLimitsResponse: TradingLimitsResponse;
  baseAsset: string;
  quoteAsset: string;
};

const parseOpenDEXassets = ({
  balanceResponse,
  tradingLimitsResponse,
  baseAsset,
  quoteAsset,
}: ParseOpenDEXassetsParams): OpenDEXassetAllocation => {
  const balancesMap = balanceResponse.getBalancesMap();
  const baseAssetBalances = balancesMap.get(baseAsset);
  if (!baseAssetBalances) {
    throw errors.BALANCE_MISSING(baseAsset);
  }
  const baseAssetBalance = new BigNumber(
    satsToCoinsStr(baseAssetBalances.getChannelBalance())
  );
  const quoteAssetBalances = balancesMap.get(quoteAsset);
  if (!quoteAssetBalances) {
    throw errors.BALANCE_MISSING(quoteAsset);
  }
  const quoteAssetBalance = new BigNumber(
    satsToCoinsStr(quoteAssetBalances.getChannelBalance())
  );
  const tradingLimitsMap = tradingLimitsResponse.getLimitsMap();
  const baseAssetLimits = tradingLimitsMap.get(baseAsset);
  if (!baseAssetLimits) {
    throw errors.TRADING_LIMITS_MISSING(baseAsset);
  }
  const baseAssetMaxOutbound = new BigNumber(
    satsToCoinsStr(baseAssetLimits.getMaxSell())
  ).plus(new BigNumber(satsToCoinsStr(baseAssetLimits.getReservedOutbound())));
  const baseAssetMaxInbound = new BigNumber(
    satsToCoinsStr(baseAssetLimits.getMaxBuy())
  ).plus(new BigNumber(satsToCoinsStr(baseAssetLimits.getReservedInbound())));
  const quoteAssetLimits = tradingLimitsMap.get(quoteAsset);
  if (!quoteAssetLimits) {
    throw errors.TRADING_LIMITS_MISSING(quoteAsset);
  }
  const quoteAssetMaxOutbound = new BigNumber(
    satsToCoinsStr(quoteAssetLimits.getMaxSell())
  ).plus(new BigNumber(satsToCoinsStr(quoteAssetLimits.getReservedOutbound())));
  const quoteAssetMaxInbound = new BigNumber(
    satsToCoinsStr(quoteAssetLimits.getMaxBuy())
  ).plus(new BigNumber(satsToCoinsStr(quoteAssetLimits.getReservedInbound())));
  return {
    baseAssetBalance,
    quoteAssetBalance,
    baseAssetMaxOutbound,
    baseAssetMaxInbound,
    quoteAssetMaxInbound,
    quoteAssetMaxOutbound,
  };
};

export {
  parseOpenDEXassets,
  logAssetBalance,
  LogAssetBalanceParams,
  ParseOpenDEXassetsParams,
};
