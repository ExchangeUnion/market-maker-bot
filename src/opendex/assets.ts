import { combineLatest, interval, Observable } from 'rxjs';
import { map, mergeMap, repeatWhen, take, tap } from 'rxjs/operators';
import { Config } from '../config';
import { Logger } from '../logger';
import { XudClient } from '../proto/xudrpc_grpc_pb';
import { GetBalanceResponse, TradingLimitsResponse } from '../proto/xudrpc_pb';
import { OpenDEXassetAllocation } from '../trade/info';
import {
  LogAssetBalanceParams,
  ParseOpenDEXassetsParams,
} from './assets-utils';

type GetOpenDEXassetsParams = {
  config: Config;
  logger: Logger;
  logBalance: ({ logger, assets }: LogAssetBalanceParams) => void;
  xudClient$: (config: Config) => Observable<XudClient>;
  xudBalance$: (client: XudClient) => Observable<GetBalanceResponse>;
  xudTradingLimits$: (client: XudClient) => Observable<TradingLimitsResponse>;
  parseOpenDEXassets: ({
    balanceResponse,
    tradingLimitsResponse,
    quoteAsset,
    baseAsset,
  }: ParseOpenDEXassetsParams) => OpenDEXassetAllocation;
};

const getOpenDEXassets$ = ({
  config,
  logger,
  logBalance,
  xudClient$,
  xudBalance$,
  xudTradingLimits$,
  parseOpenDEXassets,
}: GetOpenDEXassetsParams): Observable<OpenDEXassetAllocation> => {
  return xudClient$(config).pipe(
    mergeMap(client => {
      return combineLatest(xudBalance$(client), xudTradingLimits$(client));
    }),
    map(([balanceResponse, tradingLimitsResponse]) => {
      return parseOpenDEXassets({
        balanceResponse,
        tradingLimitsResponse,
        quoteAsset: config.QUOTEASSET,
        baseAsset: config.BASEASSET,
      });
    }),
    tap(assets => logBalance({ assets, logger })),
    take(1),
    // refresh assets every 30 seconds
    repeatWhen(() => interval(30000))
  );
};

export { getOpenDEXassets$ };
