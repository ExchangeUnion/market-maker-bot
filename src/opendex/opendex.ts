import { Observable, of } from 'rxjs';
import { tap, map, take, mergeMap, delay } from 'rxjs/operators';
import { TradeInfo, ExchangeAssetAllocation } from '../trade/manager';
import { BigNumber } from 'bignumber.js';
import { GetBalanceResponse } from '../broker/opendex/proto/xudrpc_pb';
import { XudClient } from '../broker/opendex/proto/xudrpc_grpc_pb';
import { Config } from '../config';
import { Logger } from '../logger';

type LogAssetBalanceParams = {
  logger: Logger
  assetBalance: ExchangeAssetAllocation
};

const logAssetBalance = (
  { logger, assetBalance }: LogAssetBalanceParams
): void => {
  const { baseAssetBalance, quoteAssetBalance } = assetBalance;
  logger.trace(`Base asset balance ${baseAssetBalance.toFixed()} and quote asset balance ${quoteAssetBalance.toFixed()}}`)
};

const xudBalanceToExchangeAssetAllocation =
  (
    {
      balanceResponse,
      baseAsset,
      quoteAsset,
    }:
    {
      balanceResponse: GetBalanceResponse
      baseAsset: string,
      quoteAsset: string
    }
  ): ExchangeAssetAllocation => {
    const balancesMap = balanceResponse.getBalancesMap();
    try {
      const baseAssetBalance = new BigNumber(
        balancesMap
        .get(baseAsset)
        .getChannelBalance()
      );
      const quoteAssetBalance = new BigNumber(
        balancesMap
        .get(quoteAsset)
        .getChannelBalance()
      );
      return {
        baseAssetBalance,
        quoteAssetBalance,
      }
    } catch(e) {
      throw new Error(`OpenDEX balance does not include balance for base asset ${baseAsset} or quote asset ${quoteAsset}.`);
    }
}

const getOpenDEXassets$ = (
  {
    config,
    logger,
    logBalance,
    xudClient$,
    xudBalance$,
    xudBalanceToExchangeAssetAllocation,
  }:
  {
    config: Config
    logger: Logger
    logBalance: (
      { logger, assetBalance }: LogAssetBalanceParams
    ) => void
    xudClient$: (config: Config) => Observable<XudClient>
    xudBalance$: (client: XudClient) => Observable<GetBalanceResponse>
    xudBalanceToExchangeAssetAllocation: (
      {
        balanceResponse,
        quoteAsset,
        baseAsset,
      }:
      {
        balanceResponse: GetBalanceResponse,
        quoteAsset: string,
        baseAsset: string
      }
    ) => ExchangeAssetAllocation,
  }
): Observable<ExchangeAssetAllocation> => {
  return xudClient$(config).pipe(
    mergeMap((client) => xudBalance$(client)),
    map((balanceResponse) => {
      return xudBalanceToExchangeAssetAllocation({
        balanceResponse,
        quoteAsset: config.QUOTEASSET,
        baseAsset: config.BASEASSET,
      });
    }),
    tap((assetBalance) => logBalance({ assetBalance, logger })),
    take(1),
  )
};

const getOpenDEXorders$ = (config: Config, tradeInfo: TradeInfo): Observable<boolean> => {
  // mock implementation
  return of(true).pipe(delay(3000));
}

const getOpenDEXorderFilled$ = (config: Config): Observable<boolean> => {
  // mock implementation
  return of(true).pipe(delay(3000));
}

export {
  getOpenDEXassets$,
  xudBalanceToExchangeAssetAllocation,
  logAssetBalance,
  getOpenDEXorders$,
  getOpenDEXorderFilled$,
};
