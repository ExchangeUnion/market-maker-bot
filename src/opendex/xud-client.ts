import { Subscriber, Observable, of } from 'rxjs';
import { Config } from '../config';
import { XudClient } from '@proto/xudrpc_grpc_pb';
import fs from 'fs';
import { credentials } from '@grpc/grpc-js';
import {
  GetBalanceRequest,
  GetBalanceResponse,
} from '@proto/xudrpc_pb';
import { ServiceError } from '@grpc/grpc-js';

const getXudClient$ = (config: Config): Observable<XudClient> => {
  const cert = fs.readFileSync(config.OPENDEX_CERT_PATH);
  const sslCredentials = credentials.createSsl(cert);
  const options = {
    'grpc.ssl_target_name_override' : 'localhost',
    'grpc.default_authority': 'localhost'
  };
  const client = new XudClient(
    `${config.OPENDEX_RPC_HOST}:${config.OPENDEX_RPC_PORT}`,
    sslCredentials,
    options,
  );
  return of(client);
};

const processResponse = (subscriber: Subscriber<unknown>) => {
  return (error: ServiceError | null, response: any) => {
    if (error) {
      subscriber.error(error);
    } else {
      subscriber.next(response);
    }
  };
};

const getXudBalance$ = (client: XudClient): Observable<GetBalanceResponse> => {
  const request = new GetBalanceRequest();
  const balance$ = new Observable((subscriber) => {
    client.getBalance(request, processResponse(subscriber));
  });
  return balance$ as Observable<GetBalanceResponse>;
};

export {
  getXudClient$,
  getXudBalance$,
  processResponse,
};
