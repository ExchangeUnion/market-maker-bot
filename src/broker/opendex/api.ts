import {
  ExchangeAPI,
  Balance,
  BalanceProperties,
  CancelOrderRequest,
  QueryOrderResponse,
} from '../api';
import { Logger } from '../../logger';
import { XudGrpcClient } from './xud-client';
import { v4 as uuidv4 } from 'uuid';
import {
  OrderSide as XudOrderSide,
  PlaceOrderResponse,
  SwapSuccess,
 } from './proto/xudrpc_pb';
import { ClientReadableStream, status } from '@grpc/grpc-js';
import { OrderSide } from '../../enums';
import {
  MarketOrderRequest,
  LimitOrderRequest,
  StopLimitOrderRequest,
} from '../exchange';
import { OpenDexOrder } from './order';
import { satsToCoinsStr, coinsToSats } from '../../utils';

class OpenDexAPI extends ExchangeAPI {
  private logger: Logger;
  private certPath: string;
  private rpchost: string;
  private rpcport: number;
  private xudClient!: XudGrpcClient;
  private swapSubscriptions = new Map<string, (swapSuccess: SwapSuccess.AsObject) => unknown >();
  private checkConnectionTimeout: ReturnType<typeof setTimeout> | undefined;

  private swapsCompleteSubscription: ClientReadableStream<SwapSuccess> | undefined;

  constructor(
    { logger, certPath, rpchost, rpcport }:
    {
      logger: Logger,
      certPath: string,
      rpchost: string,
      rpcport: number,
    },
  ) {
    super();
    this.logger = logger;
    this.certPath = certPath;
    this.rpchost = rpchost;
    this.rpcport = rpcport;
  }

  public start = async () => {
    this.xudClient = new XudGrpcClient({
      tlscertpath: this.certPath,
      rpchost: this.rpchost,
      rpcport: this.rpcport,
    });
    this.xudClient.start();
    await this.waitForConnection();
    this.swapsCompleteSubscription = await this.xudClient.subscribeSwaps();
    this.swapsCompleteSubscription.on('data', this.onSwapComplete.bind(this));
    this.swapsCompleteSubscription.on('end', () => {
      this.start();
      // TODO: cancel all orders on other exchanges
    });
    this.swapsCompleteSubscription.on('error', (error: any) => {
      if (error.code && error.code === status.CANCELLED) {
        return;
      } else {
        this.logger.error(`failed to subscribe to xud swaps: ${error}`);
      }
    });
    this.logger.info('OpenDEX API started');
  }

  private waitForConnection = () => {
    return new Promise((resolve, reject) => {
      const verifyConnection = async () => {
        try {
          await this.xudClient.getBalance();
          resolve();
        } catch (e) {
          if (e.code === 14) {
            this.logger.warn('could not verify connection to xud, retrying in 5 sec...');
            this.checkConnectionTimeout = setTimeout(verifyConnection, 5000);
          } else {
            reject();
          }
        }
      };
      verifyConnection();
    });
  }

  public stop = async () => {
    if (this.checkConnectionTimeout) {
      clearTimeout(this.checkConnectionTimeout);
    }
    if (this.swapsCompleteSubscription) {
      this.swapsCompleteSubscription.cancel();
    }
  }

  public getAssets = async (): Promise<Balance[]> => {
    const getBalanceResponse = await this.xudClient.getBalance();
    const assetsMap = new Map<string, BalanceProperties>();
    getBalanceResponse.balancesMap.forEach((balance) => {
      const asset = balance[0];
      const free = parseFloat(
        satsToCoinsStr(balance[1].channelBalance),
      );
      const locked = parseFloat(
        satsToCoinsStr(balance[1].walletBalance),
      );
      assetsMap.set(asset, {
        free,
        locked,
      });
    });
    const tradingLimitsResponse = await this.xudClient.tradingLimits();
    tradingLimitsResponse.limitsMap.forEach((limit) => {
      const asset = limit[0];
      const maxbuy = limit[1].maxbuy;
      const maxsell = limit[1].maxsell;
      let balance = assetsMap.get(asset);
      if (balance) {
        balance = {
          maxbuy,
          maxsell,
          ...balance,
        };
        assetsMap.set(asset, balance);
      }
    });
    const balances: Balance[] = [];
    assetsMap.forEach((balance, asset) => {
      balances.push({
        asset,
        ...balance,
      });
    });
    return balances;
  }

  public newOrder = (
    orderRequest: MarketOrderRequest | LimitOrderRequest | StopLimitOrderRequest,
  ): OpenDexOrder => {
    const orderId = uuidv4();
    const order = new OpenDexOrder({
      ...orderRequest,
      orderId,
      api: this,
      logger: this.logger,
    });
    return order;
  }

  public startOrder = async (
    orderRequest: MarketOrderRequest | LimitOrderRequest | StopLimitOrderRequest,
  ): Promise<PlaceOrderResponse.AsObject> => {
    const pairId = `${orderRequest.baseAsset}/${orderRequest.quoteAsset}`;
    const quantityInSatoshis = coinsToSats(orderRequest.quantity);
    const xudOrderSide = orderRequest.orderSide === OrderSide.Buy
      ? XudOrderSide.BUY
      : XudOrderSide.SELL;
    const xudOrderRequest = {
      pairId,
      orderId: orderRequest.orderId,
      price: orderRequest.price,
      quantity: quantityInSatoshis,
      orderSide: xudOrderSide,
    };
    const placeOrderResponse = await this.xudClient.newOrder(xudOrderRequest);
    return placeOrderResponse;
  }

  public cancelOrder = async (cancelOrderRequest: CancelOrderRequest): Promise<boolean> => {
    try {
      await this.xudClient.removeOrder(cancelOrderRequest.orderId);
      return true;
    } catch (e) {
      return false;
    }
  }

  public subscribeSwap = (id: string, cb: (swapSuccess: SwapSuccess.AsObject) => unknown) => {
    this.swapSubscriptions.set(id, cb);
  }

  public queryOrder = async (): Promise<QueryOrderResponse> => {
    throw new Error('not implemented');
  }

  private onSwapComplete = (swapSuccess: SwapSuccess) => {
    const swapSuccessObject = swapSuccess.toObject();
    const swapSubscription = this.swapSubscriptions.get(swapSuccessObject.localId);
    swapSubscription && swapSubscription(swapSuccessObject);
  }

}

export { OpenDexAPI };
