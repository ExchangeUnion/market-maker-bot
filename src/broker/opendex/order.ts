import {
  MarketOrderRequest,
  LimitOrderRequest,
} from '../exchange';
import {
  OrderType,
  OrderStatus,
} from '../../enums';
import { Order } from '../order';
import {
  SwapSuccess,
 } from './proto/xudrpc_pb';
import { OpenDexAPI } from './api';

class OpenDexOrder extends Order {

  public start = async () => {
    if (this.orderType === OrderType.StopLimit) {
      throw new Error('OpenDEX broker does not currently support StopLimit orders');
    }
    const orderRequest: MarketOrderRequest = {
      orderId: this.orderId,
      baseAsset: this.baseAsset,
      quoteAsset: this.quoteAsset,
      orderSide: this.orderSide,
      orderType: this.orderType,
      quantity: this.quantity,
      price: 'mkt',
    };
    if (this.orderType === OrderType.Limit) {
      if (!this.price) {
        throw new Error('Price is required for limit orders.');
      }
      if (typeof this.price !== 'number') {
        throw new Error('Price must be a number.');
      }
      (orderRequest as unknown as LimitOrderRequest).price = this.price;
    }
    if (this.orderType === OrderType.Market) {
      (orderRequest as MarketOrderRequest).price = 'mkt';
    }
    try {
      (this.api as OpenDexAPI).subscribeSwap(
        this.orderId, this.onSwapComplete.bind(this),
      );
      await (this.api as OpenDexAPI).startOrder(orderRequest);
      this.status = OrderStatus.New;
      this.emit('status', this.status);
    } catch (e) {
      this.logger.info(`failed to start the order: ${JSON.stringify(orderRequest)}`);
      this.emit('failure', e);
    }
  }

  private onSwapComplete = (swapSuccess: SwapSuccess.AsObject) => {
    this.logger.info(`swapSuccess: ${JSON.stringify(swapSuccess)}`);
    // TODO: emit fill instead of complete when quantity is partial
    this.emit('complete', this.orderId, swapSuccess.quantity);
  }

  public cancel = async () => {
    const cancelSuccess = await this.api.cancelOrder({
      tradingPair: this.tradingPair,
      orderId: this.orderId,
    });
    if (cancelSuccess) {
      this.logger.info(`canceled order with id: ${this.orderId}`);
    }
    this.status = OrderStatus.Canceled;
    this.emit('status', OrderStatus.Canceled);
  }

}

export { OpenDexOrder };
