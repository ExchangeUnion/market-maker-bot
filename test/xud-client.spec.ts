import { XudGrpcClient } from '../src/broker/opendex/xud-client';
import { homedir } from 'os';
import {
  OrderSide as XudOrderSide,
  // PlaceOrderEvent,
} from '../src/broker/opendex/proto/xudrpc_pb';
import { v4 as uuidv4 } from 'uuid';

// This test suite is disabled by default because it requires xud to be up and running.
// This is mainly used for development proposes as the api spec deals with mocked responses.
describe.skip('xud client', () => {
  const tlscertpath = `${homedir()}/.xud/tls.cert`;
  // const tlscertpath = `${homedir()}/xud-simnet/xud-wd/tls.cert`;
  const rpchost = 'localhost';
  const rpcport = 8886;
  let xudClient: XudGrpcClient;

  beforeEach(() => {
    xudClient = new XudGrpcClient({
      tlscertpath,
      rpchost,
      rpcport,
    });
    xudClient.start();
  });

  describe('subscribeSwaps', () => {

    test('subscribes to swap complete and failed events', async () => {
      expect.assertions(1);
      const subscribeSwapsSubscription = xudClient.subscribeSwaps();
      expect(subscribeSwapsSubscription).toBeTruthy();
    });

  });

  describe('limit order', () => {

    test('creates order and cancels it', async () => {
      expect.assertions(3);
      const pairId = 'LTC/BTC';
      const quantity = 100000000;
      const orderSide = XudOrderSide.BUY;
      const price = 4000;
      const orderId = uuidv4();
      const newOrderSubscription = await xudClient.newOrder({
        pairId,
        quantity,
        orderSide,
        price,
        orderId,
      });
      const { remainingOrder } = newOrderSubscription;
      expect(remainingOrder).toBeTruthy();
      expect(remainingOrder).toEqual(
        expect.objectContaining({
          price,
          quantity,
          pairId,
          localId: orderId,
        }),
      );
      const removeOrderResponseObj = await xudClient.removeOrder(orderId);
      expect(removeOrderResponseObj).toBeTruthy();
    });

  });

  describe('market order', () => {

    test('creates order', async () => {
      expect.assertions(4);
      const pairId = 'LTC/BTC';
      const quantity = 1000;
      const orderSide = XudOrderSide.SELL;
      const price = 'mkt';
      const orderId = uuidv4();
      const newOrderSubscription = await xudClient.newOrder({
        pairId,
        quantity,
        orderSide,
        price,
        orderId,
      });
      const {
        internalMatchesList,
        swapSuccessesList,
        remainingOrder,
        swapFailuresList,
      } = newOrderSubscription;
      expect(remainingOrder).toBeUndefined();
      expect(swapSuccessesList.length).toEqual(0);
      expect(swapFailuresList.length).toEqual(0);
      expect(internalMatchesList.length).toEqual(0);
    });

  });

  describe('getbalance', () => {

    test('succeeds', async () => {
      await expect(
        xudClient.getBalance(),
      ).resolves.toBeTruthy();
    });

    test('fails', async () => {
      xudClient['rpchost'] = '123.456.789.321';
      xudClient.start();
      await expect(
        xudClient.getBalance(),
      ).rejects.toMatchSnapshot();
    });

  });

  describe('trading limits', () => {

    test('succeeds', async () => {
      await expect(
        xudClient.tradingLimits(),
      ).resolves.toEqual(
        expect.objectContaining({
          limitsMap: expect.any(Array),
        }),
      );
    });

  });

});
