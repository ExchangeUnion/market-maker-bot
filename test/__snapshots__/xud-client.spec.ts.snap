// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xud client getbalance fails 1`] = `[Error: 14 UNAVAILABLE: DNS resolution failed]`;
